Unable to postMessage a WebAssembly module to a worklet

cdumez opened the pull request:
#### a9d61c9d9e2e1e146eb0165e25df76a326997790
<pre>
Unable to postMessage a WebAssembly module to a worklet
<a href="https://bugs.webkit.org/show_bug.cgi?id=220038">https://bugs.webkit.org/show_bug.cgi?id=220038</a>
rdar://72596556

Reviewed by Darin Adler.

We were previously unable to send WASM modules (and WASM memory) via message
ports, even if the destination was in the same process (e.g. another window on
the same page or a dedicated worker).

There were two reasons for this:
- MessagePort::postMessage() was using the default SerializationContext, which
  doesn&apos;t allow encoding on WASM types. I updated the serialization context to
  address that.
- Such WASM types are not serializable over IPC and despite the changes in
  255948@main, messages would often still go over IPC, even when the
  destination is in the same process. To address this, I now mark MessagePorts
  as out-of-process when they get sent over IPC, instead of doing it when they
  get disentangled (which can happen when sending the port to an in-process
  dedicated worker)

* LayoutTests/fast/storage/serialized-script-value.html:
* LayoutTests/http/wpt/webaudio/the-audio-api/the-audioworklet-interface/audioworklet-postmessage-wasm-module.https-expected.txt: Added.
* LayoutTests/http/wpt/webaudio/the-audio-api/the-audioworklet-interface/audioworklet-postmessage-wasm-module.https.html: Added.
* LayoutTests/http/wpt/webaudio/the-audio-api/the-audioworklet-interface/processors/wasm-module-processor.js: Added.
(WASMModuleProcessor):
(WASMModuleProcessor.prototype.testModule):
(WASMModuleProcessor.prototype.handleMessage):
(WASMModuleProcessor.prototype.handleMessageError):
(WASMModuleProcessor.prototype.process):
* LayoutTests/imported/w3c/web-platform-tests/wasm/serialization/module/window-messagechannel-success-expected.txt:
* Source/WebCore/bindings/js/JSDOMGlobalObject.cpp:
(WebCore::JSDOMGlobalObject::defaultAgentClusterID):
(WebCore::JSDOMGlobalObject::agentClusterID const):
* Source/WebCore/bindings/js/JSDOMGlobalObject.h:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::agentClusterIDFromGlobalObject):
(WebCore::CloneSerializer::dumpIfTerminal):
(WebCore::CloneDeserializer::readTerminal):
* Source/WebCore/dom/MessagePort.cpp:
(WebCore::MessagePort::postMessage):
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.cpp:
(WebKit::WebSWClientConnection::postMessageToServiceWorker):
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::postMessageToServiceWorkerClient):
* Source/WebKit/WebProcess/WebCoreSupport/WebMessagePortChannelProvider.cpp:
(WebKit::WebMessagePortChannelProvider::entangleLocalPortInThisProcessToRemote):
(WebKit::WebMessagePortChannelProvider::messagePortDisentangled):
(WebKit::WebMessagePortChannelProvider::messagePortSentToRemote):
(WebKit::WebMessagePortChannelProvider::postMessageToRemote):
* Source/WebKit/WebProcess/WebCoreSupport/WebMessagePortChannelProvider.h:

Canonical link: <a href="https://commits.webkit.org/256853@main">https://commits.webkit.org/256853@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/284a06ecc460f3c2ca54bc79e59f95c5ec67e699

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96968 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6235 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30081 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106499 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166784 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100943 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6463 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34972 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89362 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103195 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102639 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83585 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31876 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/86702 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88567 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/74767 "Found 1 new API test failure: /WebKitGTK/TestUIClient:/webkit/WebKitWebView/mediaKeySystem-permission-requests (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/269 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20046 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/254 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21473 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4743 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5045 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43987 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1484 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40774 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/wasm/serialization/module/broadcastchannel-success-and-failure-expected.txt:
```diff
@@ -1,10 +1,5 @@
-TIMEOUT WebAssembly.Module cannot cross agent clusters, BroadcastChannel edition Test timed out
-
-Harness Error (FAIL), message = Unhandled rejection: The object can not be cloned.
+Harness Error (TIMEOUT), message = null
```
The failure tests cases are still failing. This is because they rely on the `messageerror` event which we don't support. When deserialization fails, we fire a regular `message` event  but with `data` being null.

darinadler replied:
Do you plan on adding `messageerror`?

cdumez replied:
Yes, we need to add it. I have a work-in-progress patch for it but haven't gotten around to finish it yet.

cdumez replied:
I have verified that with the `messageerror` event implemented, this test and the one below start passing.
My patch to support the `messageerror` event is almost ready.

darinadler commented on Source/WebCore/bindings/js/JSDOMGlobalObject.h:
```diff
@@ -31,6 +31,7 @@
 #include <JavaScriptCore/JSGlobalObject.h>
 #include <JavaScriptCore/JSObjectInlines.h>
 #include <JavaScriptCore/WeakGCMap.h>
+#include <wtf/text/WTFString.h>
```
A forward declaration usually suffices to just use something as a return type.

cdumez committed:
Unable to postMessage a WebAssembly module to a worklet
https://bugs.webkit.org/show_bug.cgi?id=220038
rdar://72596556

Reviewed by Darin Adler.

We were previously unable to send WASM modules (and WASM memory) via message
ports, even if the destination was in the same process (e.g. another window on
the same page or a dedicated worker).

There were two reasons for this:
- MessagePort::postMessage() was using the default SerializationContext, which
  doesn't allow encoding on WASM types. I updated the serialization context to
  address that.
- Such WASM types are not serializable over IPC and despite the changes in
  255948@main, messages would often still go over IPC, even when the
  destination is in the same process. To address this, I now mark MessagePorts
  as out-of-process when they get sent over IPC, instead of doing it when they
  get disentangled (which can happen when sending the port to an in-process
  dedicated worker)

* LayoutTests/fast/storage/serialized-script-value.html:
* LayoutTests/http/wpt/webaudio/the-audio-api/the-audioworklet-interface/audioworklet-postmessage-wasm-module.https-expected.txt: Added.
* LayoutTests/http/wpt/webaudio/the-audio-api/the-audioworklet-interface/audioworklet-postmessage-wasm-module.https.html: Added.
* LayoutTests/http/wpt/webaudio/the-audio-api/the-audioworklet-interface/processors/wasm-module-processor.js: Added.
(WASMModuleProcessor):
(WASMModuleProcessor.prototype.testModule):
(WASMModuleProcessor.prototype.handleMessage):
(WASMModuleProcessor.prototype.handleMessageError):
(WASMModuleProcessor.prototype.process):
* LayoutTests/imported/w3c/web-platform-tests/wasm/serialization/module/window-messagechannel-success-expected.txt:
* Source/WebCore/bindings/js/JSDOMGlobalObject.cpp:
(WebCore::JSDOMGlobalObject::defaultAgentClusterID):
(WebCore::JSDOMGlobalObject::agentClusterID const):
* Source/WebCore/bindings/js/JSDOMGlobalObject.h:
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::agentClusterIDFromGlobalObject):
(WebCore::CloneSerializer::dumpIfTerminal):
(WebCore::CloneDeserializer::readTerminal):
* Source/WebCore/dom/MessagePort.cpp:
(WebCore::MessagePort::postMessage):
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.cpp:
(WebKit::WebSWClientConnection::postMessageToServiceWorker):
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::postMessageToServiceWorkerClient):
* Source/WebKit/WebProcess/WebCoreSupport/WebMessagePortChannelProvider.cpp:
(WebKit::WebMessagePortChannelProvider::entangleLocalPortInThisProcessToRemote):
(WebKit::WebMessagePortChannelProvider::messagePortDisentangled):
(WebKit::WebMessagePortChannelProvider::messagePortSentToRemote):
(WebKit::WebMessagePortChannelProvider::postMessageToRemote):
* Source/WebKit/WebProcess/WebCoreSupport/WebMessagePortChannelProvider.h:

Canonical link: https://commits.webkit.org/256853@main
